keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+card in a Kubernetes edge cluster. Intel provides a single resource driver that
+has parameters for setting up all of these hardware functions together as
+needed for a data flow pipeline.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Scheduling is likely to be slower when many Pods request the new
+resource types, both because scheduling such a Pod involves more
+round-trips through the API server for ResourceClaimStatus updates and
+because scheduling one Pod may affect other Pods in ways that cannot
+be anticipated by the kube-scheduler. When many Pods compete for
+limited resources, multiple attempts may be needed before a suitable
+node is found.
+
+The hardware that is expected to need this more flexible allocation
+approach is going to be used by only a small subset of the pods in the
+cluster and those pods are likely to run for extended periods of time,
+so this is not a major concern.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Implementation
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which creates 
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and to call a resource kubelet plugin. That plugin returns CDI device ID(s)
+  which then must be passed to the container runtime.
+
+For a resource driver the following components are needed:
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching ResourceClaims and updating their status once it is done with
+  allocation.
```

Comment:
...May run in the cluster or not.  

?

Reply:
Correct. Added:

> It may run inside the cluster or outside of it. The only
  hard requirement is that it can connect to the API server. Optionally,
  it may also be configured as a [special scheduler extender](#kube-scheduler).